mesh/mesh.go

```diff
@@ -641,45 +641,56 @@ func (msh *Mesh) AddTXsFromProposal(ctx context.Context, layerID types.LayerID,
 
 // AddBallot to the mesh.
 func (msh *Mesh) AddBallot(ballot *types.Ballot) error {
-	if err := msh.addBallot(ballot); err != nil {
-		return err
+	decoded, err := msh.trtl.DecodeBallot(ballot)
```

Comment:
i really don't think the code should only save the ballot to DB if this call return successfully. it is conditional on tortoise's internal state and will make ballot validation flow inconsistent.

why can't a ballot be saved when it's syntactically correct?

Reply:
i also noticed that some validation in the handler is incorrect (all of validation in checkVotesConsistency is incorrect, since it doesn't decode votes, so that can be bypassed easily). so i think only tortoise can do such validation properly

Reply:
> opinion hash validation is a part of the validation.
> all of validation in checkVotesConsistency is incorrect

in that case, why not move the decoding to the gossip handler and update the validation logic there? 
since this is how you determine whether a ballot is syntactically valid or not, it make more sense to do all the necessary decoding there and, only when it's valid, call `mesh.AddBallot()`

Reply:
sorry i dont know how to do that, decoding is quite nuanced and coupled with tortoise implementation

Reply:
> sorry i dont know how to do that, decoding is quite nuanced and coupled with tortoise implementation
moving it elsewhere from the tortoise would require keeping same stuff in memory as i keep in the tortoise

i mean to expose system.Tortoise interface to the ballot gossip handler. not to move decode ballot outside of tortoise.